engine/execution/ingestion/stop_control.go

```diff
@@ -0,0 +1,167 @@
+package ingestion
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/engine/execution/state"
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// StopControl is a specialized component used by ingestion.Engine to encapsulate
+// control of pausing/stopping blocks execution.
+type StopControl struct {
+	sync.RWMutex
+	height             uint64
+	crash              bool
+	commenced          bool // if stopping process has started. We disallow any changes then.
+	set                bool // whether stop at height has been set at all, it its envisioned most of the time it won't
+	paused             bool // if execution is paused
+	stopAfterExecuting flow.Identifier
+	log                zerolog.Logger
+}
+
+// NewStopControl creates new empty NewStopControl
+func NewStopControl(log zerolog.Logger, paused bool) *StopControl {
+	return &StopControl{
+		paused:    paused,
+		commenced: paused,
+		log:       log,
+	}
+}
+
+// IsPaused returns true is block execution has been paused
+func (s *StopControl) IsPaused() bool {
+	s.RLock()
+	defer s.RUnlock()
+	return s.paused
+}
+
+// SetStopHeight sets new stop height and crash mode, and return old values:
+//   - set, whether values were previously set
+//   - height
+//   - crash
+//
+// Returns error is the stopping process has already commenced, new values will be rejected.
+func (s *StopControl) SetStopHeight(height uint64, crash bool) (bool, uint64, bool, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	oldSet := s.set
+	oldHeight := s.height
+	oldCrash := s.crash
+
+	if s.commenced {
+		return oldSet, oldHeight, oldCrash, fmt.Errorf("cannot update stop height, stopping already in progress for height %d with crash=%t", oldHeight, oldCrash)
+	}
+
+	s.set = true
```

Comment:
please log every state changes, in case we need to debug, it's better to have all logs for how we arrive at a certain state. 

we could log all state variables, if state transition happens

Reply:
Good idea, added